src/core/protocol-authorization.ts

```diff
@@ -37,24 +57,28 @@ export class ProtocolAuthorization {
 
     // get the rule set for the inbound message
     const inboundMessageRuleSet = ProtocolAuthorization.getRuleSet(
-      recordsWrite.message,
       protocolDefinition,
-      ancestorMessageChain,
+      [...ancestorMessageChain, recordsWrite.message],
       recordSchemaToLabelMap
     );
 
+    // For non-writes `recordsWrite` is actually part of the existing ancestor chain
+    if (!isWrite) {
+      ancestorMessageChain = [...ancestorMessageChain, recordsWrite.message];
+    }
```

Comment:
Not liking this, because it means `constructAncestorMessageChain()` is kinda a lie 'coz it actually doesn't construct the ancestor message chain most of the time (assuming reads happen more often than write), so it's confusing.

Reply:
Added a commit refactoring to make an "honest" `constructAncestorMessageChain()`. Took your suggestion to move the query logic into `constructAncestorMessageChain()`. Hopefully it reads cleaner now